imap/index.c

```diff
@@ -5363,161 +5319,96 @@ extractor_timeout(struct protstream *s __attribute__((unused)),
-                /* add a default timeout */
-                be->clientin = ext->clientin;
-                be->timeout = prot_addwaitevent(ext->clientin,
-                                                now + IDLE_TIMEOUT,
-                                                extractor_timeout, be);
-            }
-        }
+    if (ext->clientin) {
+        /* add a default timeout */
+        be->clientin = ext->clientin;
+        be->timeout = prot_addwaitevent(ext->clientin,
+                                        now + IDLE_TIMEOUT,
+                                        extractor_timeout, ctx);
     }
 
-    if (be->timeout && ctx->timeout) be->timeout->mark = now + ctx->timeout;
-
-    buf_free(&buf);
-
-    return r;
+    return 0;
 }
 
-
 static int extract_attachment(const char *type, const char *subtype,
                               const struct param *type_params,
                               const struct buf *data, int encoding,
                               const struct message_guid *content_guid,
                               struct getsearchtext_rock *str)
 {
-    const char *exturl =
-         config_getstring(IMAPOPT_SEARCH_ATTACHMENT_EXTRACTOR_URL);
     struct backend *be;
-    struct backend_ctx *ctx;
     struct buf buf = BUF_INITIALIZER;
     hdrcache_t hdrs = NULL;
     struct body_t body = { 0, 0, 0, 0, 0, BUF_INITIALIZER };
     const char *guidstr, *errstr = NULL;
     size_t hostlen;
+    const char **hdr, *p;
     int r = 0;
 
-    if (!str->ext) {
-        if (!exturl) return 0;
-        if (!index_text_extractor) {
-            /* This is a legitimate case for sieve and lmtpd (so we don't need
-             * to spam the logs! */
-            syslog(LOG_DEBUG, "%s: ignoring uninitialized extractor for url %s",
-                    __func__, exturl);
-            return 0;
-        }
-
-        /* cache the connection for all the attachments in this message, but only
-         * if the connection succeeds */
-        r = extractor_connect(exturl, index_text_extractor);
-        if (r) return r;
-        str->ext = index_text_extractor;
+    if (!index_text_extractor) {
+        /* This is a legitimate case for sieve and lmtpd (so we don't need
+         * to spam the logs! */
+        syslog(LOG_DEBUG, "%s: ignoring uninitialized extractor",
+                __func__);
+        return 0;
     }
 
-    struct extractor_ctx *ext = str->ext;
-
-    be = ext->be;
-    ctx = (struct backend_ctx *) be->context;
-    hostlen = strcspn(be->hostname, "/");
-
     if (message_guid_isnull(content_guid)) {
         syslog(LOG_DEBUG, "extract_attachment: ignoring null guid for %s/%s",
-                type ? type : "<null>", subtype ? subtype : "<null>");
+               type ? type : "<null>", subtype ? subtype : "<null>");
         return 0;
     }
 
-    /* Fetch previously extracted text */
-    unsigned statuscode = 0;
+    struct extractor_ctx *ext = str->ext = index_text_extractor;
+
+    r = extractor_connect(ext);
```

Comment:
This is still fine to call if we're already connected - it won't even ping any more, just check for a socket fd.